### ISSUE TITLE
LSTM model save warning

### ISSUE THREAD

**Issue by StevenHuang2020** [keras-team/keras#15964]
**System information**.
- OS Platform and Distribution: Windows10
- TensorFlow installed from: binary
- TensorFlow version : v2.7.0-rc1-69-gc256c071bb2 2.7.0
- Python version: 3.9.10

**Describe the problem**.
When model.save called, the below warning message occurred:
```
WARNING:absl:Found untraced functions such as lstm_cell_layer_call_fn, lstm_cell_layer_call_and_return_conditional_losses, lstm_cell_layer_call_fn, lstm_cell_layer_call_and_return_conditional_losses, lstm_cell_layer_call_and_return_conditional_losses while saving (showing 5 of 5). These functions will not be directly callable after loading.
WARNING:absl:<keras.layers.recurrent.LSTMCell object at 0x0000021419A6A820> has the same name 'LSTMCell' as a built-in Keras object. Consider renaming <class 'keras.layers.recurrent.LSTMCell'> to avoid naming conflicts when loading with `tf.keras.models.load_model`. If renaming is not possible, pass the object in the `custom_objects` parameter of the load function.
```

I just upgraded python from 3.6 to 3.9, reinstall Tensorflow under Python 3.9.10. I found the LSTM model hard to train to reduce loss value and the previous warning information occurred. The model trained very well with TensorFlow 2.6.2 and there are no warnings.

My model creation code is as follows:
```
from tensorflow.keras.layers import Dense, LSTM
from tensorflow.keras.models import Sequential
from tensorflow.keras import optimizers

def createModel(look_back=1):
    model = Sequential()
    model.add(LSTM(100, return_sequences=True, input_shape=(1, look_back)))
    # model.add(LSTM(100, activation='relu', return_sequences=True))
    # model.add(LSTM(50, activation='relu', return_sequences=True))
    model.add(Dense(30, activation='relu'))
    model.add(Dense(20, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # fixed learning rate
    # lr = 1e-2

    # learning rate schedule
    lr = optimizers.schedules.ExponentialDecay(initial_learning_rate=1e-5,
                                               decay_steps=10000,
                                               decay_rate=0.99)

    # opt = optimizers.SGD(learning_rate=lr)
    # opt = optimizers.SGD(learning_rate=lr, momentum=0.8, nesterov=False)
    # opt = optimizers.RMSprop(learning_rate=lr, rho=0.9, epsilon=1e-08)
    opt = optimizers.Adam(learning_rate=lr)
    # opt = optimizers.Adadelta(learning_rate=lr)
    # opt = optimizers.Adagrad(learning_rate=lr)
    # opt = optimizers.Adamax(learning_rate=lr)
    # opt = optimizers.Nadam(learning_rate=lr)
    # opt = optimizers.Ftrl(learning_rate=lr)

    model.compile(optimizer=opt, loss='mse')
    # model.summary()
    return model
```

**Comment by sushreebarsa** [comment]
@StevenHuang2020 I was able to run the code successfully on colab using TF v2.7.0 and tf-nightly(2.9.0.dev20220126)  and didn't face the warning reported here.Could you please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/997a7d67340735a857995fe3231a33f2/15964.ipynb#scrollTo=SH7P1z-zZVVr) and confirm the same?
Please provide the complete code and let us know if we are missing something to replicate the issue? Thanks!

**Comment by StevenHuang2020** [comment]
Here is my code.
https://github.com/StevenHuang2020/COVID-19-Statistics/blob/e73c17cdc14ce38616164334dde3ed7c09543328/coronavirus/predictStatistics.py

The warning message is shown when model.save called after training

![image](https://user-images.githubusercontent.com/61686583/151493297-229f2618-f805-4d4d-97f8-ccf864e7ac68.png)

**Comment by StevenHuang2020** [comment]
I simplified it to 2 files, the dataset and py file.

dataset link:
https://raw.githubusercontent.com/owid/covid-19-data/master/public/data/owid-covid-data.csv
training file:
[predictStatistics.py.txt](https://github.com/keras-team/keras/files/7956105/predictStatistics.py.txt)
You can verify the above problem with this.

**Comment by sushreebarsa** [comment]
@jvishnuvardhan Was able to replicate the issue on colab using TF v2.7.0 , please find the gist [here ](https://colab.research.google.com/gist/sushreebarsa/001c1cfa939c1ae5e385c3f2c00ba072/keras15964.ipynb) for reference.Thanks!

**Comment by jvishnuvardhan** [comment]
Similar issue in TF repository https://github.com/tensorflow/tensorflow/issues/47554.

**Comment by haifeng-jin** [comment]
Summary: unexpected long warning messages with LSTM when saving the model.
Verified with TF nightly:
https://colab.research.google.com/drive/1vuGwbaA9lXDnmuLUv3CvGONOOKllKa2D?resourcekey=0-7wqko2DJyKWPyiEPVo2YCg&usp=sharing

**Comment by haifeng-jin** [comment]
@qlzh727 @k-w-w @hertschuh

**Comment by sara-02** [comment]
Update, if I am not running any callbacks to save the model, the warning does not occur when running normal `model.fit`.

**Comment by hertschuh** [comment]
There are three separate problems described in this issue:

1. The `Found untraced functions such as ...` warning. I tested with earlier versions of Tensorflow and Keras (2.6.2 and 2.6.0) and the warning was already there. I believe it has to do with some internal management of functions but does not constitute any change between 2.6.2 and 2.7.

2. The `<keras.layers.recurrent.LSTMCell object at ...> has the same name 'LSTMCell'...` warning. This is indeed new and wasn't there in 2.6.2. This warning is a side effect of adding messaging in Keras when custom classes collide with built-in classes. This warning is not a change in the saving behavior nor a change in the behavior of the LSTM. The conflict it detects was already there in earlier versions and has to do with having multiple versions of LSTMCell within Keras. I am working on fixing this bug by changing how LSTMCell works, which will remove the warning. But to be clear, it has no other side effect besides the warning message.

3. The part about *I found the LSTM model hard to train to reduce loss value [...] The model trained very well with TensorFlow 2.6.2*. As far as I can tell, this is unrelated to the warnings and needs a separate investigation. @StevenHuang2020 any details you can provide on the change in behavior would be useful, thanks. In particular, do you reload the model at any point for training? Or are you saving it just for later use?

@sara-02 Correct, the warnings have to do with the saving only and are indeed not related to `model.fit`.

**Comment by sara-02** [comment]
> There are three separate problems described in this issue:
> 
> 1. The `Found untraced functions such as ...` warning. I tested with earlier versions of Tensorflow and Keras (2.6.2 and 2.6.0) and the warning was already there. I believe it has to do with some internal management of functions but does not constitute any change between 2.6.2 and 2.7.
> 2. The `<keras.layers.recurrent.LSTMCell object at ...> has the same name 'LSTMCell'...` warning. This is indeed new and wasn't there in 2.6.2. This warning is a side effect of adding messaging in Keras when custom classes collide with built-in classes. This warning is not a change in the saving behavior nor a change in the behavior of the LSTM. The conflict it detects was already there in earlier versions and has to do with having multiple versions of LSTMCell within Keras. I am working on fixing this bug by changing how LSTMCell works, which will remove the warning. But to be clear, it has no other side effect besides the warning message.
> 3. The part about _I found the LSTM model hard to train to reduce loss value [...] The model trained very well with TensorFlow 2.6.2_. As far as I can tell, this is unrelated to the warnings and needs a separate investigation. @StevenHuang2020 any details you can provide on the change in behavior would be useful, thanks. In particular, do you reload the model at any point for training? Or are you saving it just for later use?
> 
> @sara-02 Correct, the warnings have to do with the saving only and are indeed not related to `model.fit`.

@hertschuh I need to check this part as I was running callbacks in my model.fit and I am also using the custom CRF code. Let to fine-grain the components and revert.

**Comment by hertschuh** [comment]
@StevenHuang2020, about 3 (the fact that the model doesn't train as well on 2.7.0 as it did on 2.6.2), it turns out that this is a side effect of [this change](https://github.com/keras-team/keras/commit/cc91cef921e7ac542ff056122570e4e0b0518b21). Basically, Keras no longer automatically uprank inputs and targets if they're only 1-dimensional. On 2.7.0, it causes the model to train against targets that are the wrong shape.

Per the model summary:
```
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 lstm_1 (LSTM)               (None, 1, 100)            40800     
                                                                 
 dense_4 (Dense)             (None, 1, 30)             3030      
                                                                 
 dense_5 (Dense)             (None, 1, 20)             620       
                                                                 
 dense_6 (Dense)             (None, 1, 10)             210       
                                                                 
 dense_7 (Dense)             (None, 1, 1)              11        
                                                                 
=================================================================
```
The output is of rank 3 `(None, 1, 1)`. However, your `y_train` is one-dimensional.

If you simply uprank `y_train` in the same way you reshape your `x_train` in `prepareDataset` like this:
```py
def prepareDataset(dataset, look_back):
    ...
    X = np.reshape(X, (X.shape[0], 1, X.shape[1]))
    Y = np.reshape(X, (Y.shape[0], 1, 1))
```
Then you get similar results on 2.6.2 and 2.7.0, that is a loss in the order of 1e-6.

**Comment by hertschuh** [comment]
@StevenHuang2020 removing `return_sequences=True` makes it so that it defaults to `False`, and then the output of the LSTM layer has one less dimensions. This removes the dimensions mismatch between the output and the training targets. See the model summaries:

with `model.add(LSTM(100, return_sequences=True, input_shape=(1, look_back)))`
```
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 lstm_1 (LSTM)               (None, 1, 100)            40800     
                                                                 
 dense_4 (Dense)             (None, 1, 30)             3030      
                                                                 
 dense_5 (Dense)             (None, 1, 20)             620       
                                                                 
 dense_6 (Dense)             (None, 1, 10)             210       
                                                                 
 dense_7 (Dense)             (None, 1, 1)              11        
                                                                 
=================================================================
```

with `model.add(LSTM(100, input_shape=(1, look_back)))`
```
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 lstm_1 (LSTM)               (None, 100)               40800     
                                                                 
 dense_4 (Dense)             (None, 30)                3030      
                                                                 
 dense_5 (Dense)             (None, 20)                620       
                                                                 
 dense_6 (Dense)             (None, 10)                210       
                                                                 
 dense_7 (Dense)             (None, 1)                 11        
                                                                 
=================================================================
```

For more details on `return_sequences`, please look at the documentation for [RNN base class](https://keras.io/api/layers/recurrent_layers/rnn/), which is the base class for [LSTM](https://keras.io/api/layers/recurrent_layers/lstm/). In particular, in the **Output shape** section explains how `return_sequences` changes the shape.

Also, the behavior of `return_sequences` has not changed between 2.6 and 2.7.

**Comment by hertschuh** [comment]
#16164 fixes the following warning:
```
WARNING:absl:<keras.layers.recurrent.LSTMCell object at 0x0000021419A6A820> has the same name 'LSTMCell' as a built-in Keras object. Consider renaming <class 'keras.layers.recurrent.LSTMCell'> to avoid naming conflicts when loading with `tf.keras.models.load_model`. If renaming is not possible, pass the object in the `custom_objects` parameter of the load function.
```

Considering that this issue was primarily focused on this warning, closing now.

**Comment by jjl2** [comment]
> #16164 fixes the following warning:
> 
> ```
> WARNING:absl:<keras.layers.recurrent.LSTMCell object at 0x0000021419A6A820> has the same name 'LSTMCell' as a built-in Keras object. Consider renaming <class 'keras.layers.recurrent.LSTMCell'> to avoid naming conflicts when loading with `tf.keras.models.load_model`. If renaming is not possible, pass the object in the `custom_objects` parameter of the load function.
> ```
> 
> Considering that this issue was primarily focused on this warning, closing now.

Hi there, I just encountered this same warning in Tensorflow 2.7 when I added a ModelCheckpoint callback to an LSTM training run.  

**_I also noticed a huge performance slowdown when this warning is thrown._**  My training runs at least 20 times more slowly.   It's really bad. I get long pauses during an epoch, and a longer pause at the end of each epoch when the warning prints.  Neither my models nor my data sets are particularly large.  The saved model folder is less than 1MB, and my data is only about 10MB.

Has anyone else noticed a performance drop when encountering this error?

Have these commits been merged into a release we can try?  I'm not compiling from source.  I shudder when I recall my last attempts to get Bazel running (about two years ago).
